File: 23-Merge-k-Sorted-Lists/solution.js
/**
 * Definition for singly-linked list.
 * function ListNode(val) {
 *     this.val = val;
 *     this.next = null;
 * }
 */
/**
 * @param {ListNode[]} lists
 * @return {ListNode}
 */
var mergeKLists = function(lists) {
    int n = lists.length;
    if(n == 0) return null;
    int i = 0, j = n - 1;
    Comparator<ListNode> c = new Comparator<ListNode>() {
        public int compare(ListNode n1, ListNode n2) {
            return n1.val - n2.val;
        }  
    };
    PriorityQueue<ListNode> pq = new PriorityQueue<ListNode>(n, c);
    for(ListNode l: lists) {
        if(l != null) pq.add(l);
    }
    ListNode dummy = new ListNode(0);
    ListNode pre = dummy;
    while(!pq.isEmpty()) {
        ListNode cur = pq.poll();
        pre.next = cur;
        pre = pre.next;
        if(cur.next != null) pq.add(cur.next);
    }
    return dummy.next;
};

File: 23-Merge-k-Sorted-Lists/solution.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution(object):
    def mergeKLists(self, lists):
        """
        :type lists: List[ListNode]
        :rtype: ListNode
        """
        if not lists:
            return None
        heap = []
        for l in lists:
            if l:
                heap.append([l.val, l])
        heapq.heapify(heap)
        head = ListNode(0)
        pre = head
        while heap:
            cur = heapq.heappop(heap)[1]
            pre.next = cur
            pre = cur
            if cur.next:
                heapq.heappush(heap, [cur.next.val, cur.next])
        return head.next
        